src/test/java/com/cloudbees/jenkins/plugins/file_leak_detector/FileHandleDumpTest.java

```diff
@@ -0,0 +1,143 @@
+            assertThat("There should be a file opened by this method", getOpenDescriptors(), hasItem(allOf(
+                    containsString(leakyFile.getPath() + " by thread:"),
+                    containsString("FileHandleDumpTest.detectFileLeak(")
+            )));
+        }
+        assertThat("There should not be a file opened by this method", getOpenDescriptors(), not(hasItem(allOf(
+                containsString(leakyFile.getPath() + " by thread:"),
+                containsString("FileHandleDumpTest.detectFileLeak(")
+        ))));
+    }
+
+    @Test
+    public void detectPipeLeak() throws Exception {
+        activateFileLeakDetector();
+        Pipe p = Pipe.open();
+        try {
+            assertThat("There should be a pipe sink and source channel opened by this method", getOpenDescriptors(), allOf(
+                hasItem(allOf(
+                        containsString("Pipe Sink Channel by thread:"),
+                        containsString("FileHandleDumpTest.detectPipeLeak(")
+                )), hasItem(allOf(
+                        containsString("Pipe Source Channel by thread:"),
+                        containsString("FileHandleDumpTest.detectPipeLeak(")
+                ))));
+        } finally {
+            p.sink().close();
+            p.source().close();
+        }
+        assertThat("There should not be a pipe sink or source channel opened by this method", getOpenDescriptors(), not(anyOf(
+                hasItem(allOf(
+                        containsString("Pipe Sink Channel by thread:"),
+                        containsString("FileHandleDumpTest.detectPipeLeak(")
+                )), hasItem(allOf(
+                        containsString("Pipe Source Channel by thread:"),
+                        containsString("FileHandleDumpTest.detectPipeLeak(")
+                )))));
+    }
+
+    @Test
+    public void detectSelectorLeak() throws Exception {
+        activateFileLeakDetector();
+        try (Selector unused = Selector.open()) {
+            assertThat("There should be a selector opened by this method", getOpenDescriptors(), hasItem(allOf(
+                    containsString("selector by thread:"),
+                    containsString("FileHandleDumpTest.detectSelectorLeak(")
+            )));
+        }
+        assertThat("There should not be a selector opened by this method", getOpenDescriptors(), not(hasItem(allOf(
+                containsString("selector by thread:"),
+                containsString("FileHandleDumpTest.detectSelectorLeak(")
+        ))));
+    }
+
+    /**
+     * Activates the file-leak-detector library. Only activates once for the
+     * whole test suite, but is called in every test so they can run in any
+     * order, and cannot not be called until Jenkins starts.
+     */
+    private void activateFileLeakDetector() throws Exception {
+        FileHandleDump fileHandleDump = ManagementLink.all().get(FileHandleDump.class);
+        fileHandleDump.doActivate("");
```

Comment:
[Fails](https://ci.jenkins.io/job/Plugins/job/file-leak-detector-plugin/job/PR-4/1/testReport/com.cloudbees.jenkins.plugins.file_leak_detector/FileHandleDumpTest/detectFileLeak/) on Windows.

Reply:
Good catch. I'll investigate to see whether we can fix it here or if the library itself doesn't work on Windows.

Reply:
I really have no idea. You can always

```groovy
buildPlugin(platforms: ['linux'])
```

if it is _known_ not to work on Windows.

Reply:
Updating to JNR fixed the main issue, but pipes aren't detected correctly on Windows by the library, so I've ignored that test on Windows for now. See https://github.com/kohsuke/file-leak-detector/issues/36.